src/Workspaces/Core/Portable/Workspace/Solution/FilePathToDocumentIdsMap.cs

```diff
@@ -0,0 +1,95 @@
+    /// (because only documents whose trees have been parsed are kept out).  To keep things fast, instead of actually
+    /// dropping all those files from our <see cref="_map"/> we instead only keep track of added documents and mark that
+    /// we're frozen.  Then, when a client actually asks for the document ids in a particular solution, we only return the
+    /// actual set present in that solution instance.
+    /// </summary>
+    public readonly bool IsFrozen;
+    private readonly ImmutableDictionary<string, ImmutableArray<DocumentId>> _map;
+
+    private FilePathToDocumentIdsMap(bool isFrozen, ImmutableDictionary<string, ImmutableArray<DocumentId>> map)
+    {
+        IsFrozen = isFrozen;
+        _map = map;
+    }
+
+    public bool TryGetValue(string filePath, out ImmutableArray<DocumentId> documentIdsWithPath)
+        => _map.TryGetValue(filePath, out documentIdsWithPath);
+
+    public static bool operator ==(FilePathToDocumentIdsMap left, FilePathToDocumentIdsMap right)
+        => left.IsFrozen == right.IsFrozen && left._map == right._map;
+
+    public static bool operator !=(FilePathToDocumentIdsMap left, FilePathToDocumentIdsMap right)
+        => !(left == right);
+
+    public override int GetHashCode()
+        => throw new NotSupportedException();
+
+    public override bool Equals([NotNullWhen(true)] object? obj)
+        => obj is FilePathToDocumentIdsMap map && Equals(map);
+
+    public Builder ToBuilder()
+        => new(IsFrozen, _map.ToBuilder());
+
+    public FilePathToDocumentIdsMap ToFrozen()
+        => IsFrozen ? this : new(isFrozen: true, _map);
+
+    public readonly struct Builder
+    {
+        private readonly bool _isFrozen;
+        private readonly ImmutableDictionary<string, ImmutableArray<DocumentId>>.Builder _builder;
+
+        public Builder(bool isFrozen, ImmutableDictionary<string, ImmutableArray<DocumentId>>.Builder builder)
+        {
+            _isFrozen = isFrozen;
+            _builder = builder;
+        }
+
+        public FilePathToDocumentIdsMap ToImmutable()
+            => new(_isFrozen, _builder.ToImmutable());
+
+        public bool TryGetValue(string filePath, out ImmutableArray<DocumentId> documentIdsWithPath)
+            => _builder.TryGetValue(filePath, out documentIdsWithPath);
+
+        public void Add(string? filePath, DocumentId documentId)
+        {
+            if (RoslynString.IsNullOrEmpty(filePath))
+                return;
+
+            _builder.MultiAdd(filePath, documentId);
+        }
+
+        public void Remove(string? filePath, DocumentId documentId)
+        {
+            if (RoslynString.IsNullOrEmpty(filePath))
+                return;
+
+            if (!this.TryGetValue(filePath, out var documentIdsWithPath) || !documentIdsWithPath.Contains(documentId))
+                throw new ArgumentException($"The given documentId was not found");
+
+            _builder.MultiRemove(filePath, documentId);
```

Comment:
>MultiRemove [](http://example.com/codeflow?start=21&length=11)

Bit of a shame that the MultiRemove extension method doesn't return a bool on whether it was removed.

Also, in looking at the MultiRemove extension method, it seems odd that the Length == 1 case just removes the key from the dictionary without verifying that the only entry in the array contains the value.

Reply:
Oh ick.  That seems awful. I will see about fixing that tomorrow

Reply:
fixed that.